File: temp p3/P3/P3/TranRobot.cs
/**
 * Author:Shaun Lee
 * Platform:VSCode
 * Class Overview:This class has all the functionality of robot plus up down left right
 *
 * Class Invariant: 
 * don't go out of bounds in the grid
 * Forward() moves the position of the robot north
 * Back() move the position of the robot south
 * Left() moves the position of the robot left 
 * Right() moves the position of the robot right
 * moves moves the robot forward depending on the direction of the robot
 *
 * Interface Invariant:   
 * robot cannot go through walls
 * rCoord and cCoord cannot go below 0 
 * using the direction function changes the direction of the robot
 * 
 **/
public class TranRobot : robot
{
    public TranRobot(string str,double drainRate) : base(str,drainRate)
    {
        
    }
    /**
     *PreCondition: existing 2 ints
     *PostCondition: changing state of ints
     **/
    public void Forward()
    {
        if (upSensor.isValid(rCoord,cCoord))
        {
            nAct.MoveForward(ref rCoord, ref cCoord);
        }
        orientationState = orientation.Up;
    }
    /**
     *PreCondition: existing 2 ints
     *PostCondition: changing state of ints
     **/
    public void Left()
    {
        if (leftSensor.isValid(rCoord,cCoord))
        {
            wAct.MoveForward(ref rCoord, ref cCoord);
        }
        orientationState = orientation.Left;
    }
    /**
     *PreCondition: existing 2 ints
     *PostCondition: changing state of ints
     **/
    public void Right()
    {
        if (rightSensor.isValid(rCoord,cCoord))
        {
            eAct.MoveForward(ref rCoord, ref cCoord);
        }
        orientationState = orientation.Right;
    }
    /**
     *PreCondition: existing 2 ints
     *PostCondition: changing state of ints
     **/
    public void Back()
    {
        if (downSensor.isValid(rCoord,cCoord))
        {
            sAct.MoveForward(ref rCoord, ref cCoord);
        }    
        orientationState = orientation.Down;
    }

    /**
     *PreCondition: none
     *PostCondition: changing state of ints
     **/
    public override void Move()
    { 
        switch (orientationState)
		{
			case orientation.Up:
				while (upSensor.isValid(rCoord,cCoord))
				{
					nAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			case orientation.Down: 
				while (downSensor.isValid(rCoord,cCoord))
				{
					sAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			case orientation.Right:
				while (rightSensor.isValid(rCoord,cCoord))
				{
					eAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			case orientation.Left:
				while (leftSensor.isValid(rCoord,cCoord))
				{
					wAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			default:
				break;
		} 
    }

    /**
     *PreCondition: none
     *PostCondition: changing state of coord
     **/
    public override void MoveOne()
    {
        switch (orientationState)
		{
			case orientation.Up:
				if (upSensor.isValid(rCoord,cCoord))
				{
					nAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			case orientation.Down:
				if (downSensor.isValid(rCoord,cCoord))
				{
					sAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			case orientation.Right:
				if (rightSensor.isValid(rCoord,cCoord))
				{
					eAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			case orientation.Left:
				if (leftSensor.isValid(rCoord,cCoord))
				{
					wAct.MoveForward(ref rCoord, ref cCoord);
				}
				break;
			default:
				break;
		}
    }
}

    

/*Implementation invariant:
 * Constructor argument passes the arguments to the robot
 * moving a direction changes the direction of the robot
 * rCoord and cCoord are changed through references passed in
 * override parent's move() with transrobot()
 * directional functions uses directional actuator with directional sensors
 */

File: P4/TranRobot.cpp
#include "tranRobot.h"

TranRobot::TranRobot(const std::string str, double drainRate) : robot(str,drainRate)
{

}


void TranRobot::Forward()
{
	moveCount++;
	if (upCheck(rCoord,cCoord))
	{
		for(int i = 0; i < SIZE; i++)
		{
		nAct->MoveForward(rCoord, cCoord); 
		}
	}
	orientationState = orientation::Up;
}

void TranRobot::Left()
{
	moveCount++;
	if (leftCheck(rCoord,cCoord))
	{
		wAct->MoveForward(rCoord, cCoord);
	}
	orientationState = orientation::Left;
}

void TranRobot::Right()
{
	moveCount++;
	if (rightCheck(rCoord,cCoord))
	{
		eAct->MoveForward(rCoord, cCoord);
	}
	orientationState = orientation::Right;
}

void TranRobot::Back()
{
	moveCount++;
	if (downCheck(rCoord,cCoord))
	{
		sAct->MoveForward(rCoord, cCoord);
	}
	orientationState = orientation::Down;
}

void TranRobot::Move()
{ 
	switch (orientationState)
	{
		case orientation::Up:
			while (upCheck(rCoord,cCoord))
			{
				moveCount++;
				nAct->MoveForward(rCoord, cCoord);
			}
			break;
		case orientation::Down:
			while (downCheck(rCoord,cCoord))
			{
				moveCount++;
				sAct->MoveForward(rCoord, cCoord);
			}
			break;
		case orientation::Right:
			while (rightCheck(rCoord,cCoord))
			{
				moveCount++;
				eAct->MoveForward(rCoord, cCoord);
			}
			break;
		case orientation::Left:
			while (leftCheck(rCoord,cCoord))
			{
				moveCount++;
				wAct->MoveForward(rCoord, cCoord);
			}
			break;
		default:
			break;
	}
}

void TranRobot::MoveOne()
{
	switch (orientationState)
	{
		case orientation::Up:
			if (upCheck(rCoord,cCoord))
			{
				nAct->MoveForward(rCoord, cCoord);
			}
			break;
		case orientation::Down:
			if (downCheck(rCoord,cCoord))
			{
				sAct->MoveForward(rCoord, cCoord);
			}
			break;
		case orientation::Right:
			if (rightCheck(rCoord,cCoord))
			{
				eAct->MoveForward(rCoord, cCoord);
			}
			break;
		case orientation::Left:
			if (leftCheck(rCoord,cCoord))
			{
				wAct->MoveForward(rCoord, cCoord);
			}
			break;
		default:
			break;
	}
}

/*Implementation invariant:
 * Constructor argument passes the arguments to the robot
 * moving a direction changes the direction of the robot
 * rCoord and cCoord are changed through references passed in
 * override parent's move() with transrobot()
 * directional functions uses directional actuator with directional sensors
 * add bool function that checks if at least 2 sensors have 80% battery and returns a bool from each direction
 * default argument for drainRate is random between 1 - 10
 */